akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala

```diff
@@ -16,46 +18,64 @@ trait CodingDirectives {
   import MiscDirectives._
   import RouteDirectives._
   import CodingDirectives._
+  import HeaderDirectives._
 
   // encoding
 
   /**
-   * Wraps its inner Route with encoding support using the given Encoder.
-   */
-  def encodeResponse(encoder: Encoder): Directive0 =
-    responseEncodingAccepted(encoder.encoding) &
-      mapResponse(encoder.encode(_)) &
-      cancelRejections(classOf[UnacceptedResponseEncodingRejection])
-
-  /**
    * Rejects the request with an UnacceptedResponseEncodingRejection
    * if the given encoding is not accepted for the response.
    */
   def responseEncodingAccepted(encoding: HttpEncoding): Directive0 =
     extract(_.request.isEncodingAccepted(encoding))
-      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(encoding)))
+      .flatMap(if (_) pass else reject(UnacceptedResponseEncodingRejection(Set(encoding))))
 
   /**
-   * Wraps its inner Route with response compression, using the specified
-   * encoders in the given order of preference.
-   * If no encoders are specifically given Gzip, Deflate and NoEncoding
-   * are used in this order, depending on what the client accepts.
+   * Wraps its inner Route with response encoding, using one of the specified
+   * encoders negotiated with the encodings the client accepts.
+   *
+   * The negotiation logic is defined by http://tools.ietf.org/html/rfc7231#section-5.3.4.
    */
-  def compressResponse(encoders: Encoder*): Directive0 =
-    theseOrDefault(encoders).map(encodeResponse(_)).reduce(_ | _)
+
+  def encodeResponseWith(encoders: Encoder*): Directive0 = _encodeResponse(immutable.Seq(encoders: _*))
 
   /**
-   * Wraps its inner Route with response compression if and only if the client
-   * specifically requests compression with an `Accept-Encoding` header.
+   * Wraps its inner Route with response compression unless the client denied compression with an
+   * `Accept-Encoding` header.
    */
-  def compressResponseIfRequested(): Directive0 = compressResponse(NoCoding, Gzip, Deflate)
+  def encodeResponse: Directive0 = _encodeResponse(DefaultCoders)
```

Comment:
Mathias, I agree that the two directives are all that is needed.

For the special cases we have 3 scenarios to handle:

1. [No Accept header](https://github.com/2beaucoup/akka/blob/wip-16593-negotiate-encoding/akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala#L50-L52): That is the only case where the spec is a little vague. We agreed above that we should default to `NoCoding` here.
2. [Empty Accept header](https://github.com/2beaucoup/akka/blob/wip-16593-negotiate-encoding/akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala#L53-L55): From the spec: "An Accept-Encoding header field with a combined field-value that is empty implies that the user agent does not want any content-coding in response." Isn't this in contradiction to your "compress-if-no-other-preference" behavior? You want to compress in this case, right?
3. [Non-empty Accept header](https://github.com/2beaucoup/akka/blob/wip-16593-negotiate-encoding/akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala#L56-L65): Negotiate the supplied accepted encodings and their qVals against the supplied encoders. (I added the client supplied order as an additional preference criteria.) The spec further says: "If an Accept-Encoding header field is present in a request and none of the available representations for the response have a content-coding that is listed as acceptable, the origin server SHOULD send a response without any content-coding." I'll add impl for that part.

To sum it up from my POV:
```scala
/**
  * Encodes the response with the encoding that is requested by the client with the `Accept-
  * Encoding` header. The response encoding is determined by the rules specified in
  * http://tools.ietf.org/html/rfc7231#section-5.3.4.
  *
  * If the `Accept-Encoding` header is missing or empty or specifies an encoding other than
  * identity, gzip or deflate no encoding is used. 
  */
def encodeResponse = _encodeResponse(immutable.Seq(NoCoding, Gzip, Deflate))

/**
  * Encodes the response with the encoding that is requested by the client with the `Accept-
  * Encoding` header. The response encoding is determined by the rules specified in
  * http://tools.ietf.org/html/rfc7231#section-5.3.4.
  *
  * If the `Accept-Encoding` header is missing then the response is encoded using the `first`
  * encoder.
  *
  * If the `Accept-Encoding` header is empty and `NoCoding` is part of the encoders then no
  * response encoding is used. Otherwise the request is rejected.
  */
def encodeResponseWith(first: Encoder, more: Encoder*) = _encodeResponse(immutable.Seq(first +: more: _*))
```

Does that make sense?